### ISSUE TITLE
Recipe request: Immutable flag and how it works

### ISSUE THREAD

**Issue by kevmodrome** [svelte-society/sveltesociety.dev#66]
Having read a comment by @ryanatkn in #60 it's dawning on me that we should have a recipe on how the immutable flag works, what it does, any performance implications, etc.

**Comment by stephane-vanraes** [comment]
I was just wondering the same, also regardig the work that has recently been done for 'will it scale ?'
As far as I know the immutable components have a lower footprint, so those would scale better ?
Guess this is more of a deepdive though, maybe one of the svelte contributors would be willing to work on this

**Comment by benmccann** [comment]
I recently asked the same question and Tan Li Hau explained to me.

The file size does not decrease for immutable components, but it may offer better performance.

It only affects https://github.com/sveltejs/svelte/blob/935caee2697deb6235af74d22fda60314527452d/src/compiler/compile/render_dom/index.ts#L112

which is whether to choose `not_equal` or `safe_not_equal` inside `$$invalidate` to determine whether the variable has changed.
`immutable: false` will use `safe_not_equal`, meaning that it will always treat `$$invalidate` array or function to be always true, i.e. always mark it as dirty.

You can see the difference in https://svelte.dev/examples#immutable-data

**Comment by ryanatkn** [comment]
That's about the extent of my knowledge. That example demonstrates a compelling case to use the immutable flag when possible, because it significantly optimizes child components in a list. (likely one of the most common performance gotcha scenarios) The parent component could opt into mutability, and use mutable data structures like maps, without de-optimizing all of its children. (I think?)

Are there any cases where `$$invalidate` gets called other than reassignments and new props?

(@ryanatkn btw not ryanatk, there's several of us Ryan Atkinsons floating around, rynpatk being another lol)

**Comment by kevmodrome** [comment]
(my bad, updated the name Ryan 🙈)

**Comment by sw-yx** [comment]
I still have never used the immutable flag  would love a recipe hehe

**Comment by sw-yx** [comment]
took a first stab at the immutable recipe :) i think pretty much what was transcribed here is enough https://github.com/svelte-society/sveltesociety.dev/pull/167

**Comment by frederikhors** [comment]
I have read all the text in https://github.com/svelte-society/sveltesociety.dev/pull/167 and I wanna thank you very much, @sw-yx: it is clearer than the tutorial and the docs on the Svelte website.

But I would like there also what is NOT recommended to do, e.g.:

`"Since it is better (faster) to use the `immutable` option in lists, this is WHAT YOU SHOULD NOT DO, what is WRONG (slower) to do:" `.

Because - in reality - both from that page and from the Svelte site I have not yet understood it well, and like me other people too.

In a nutshell: we can add more concrete examples of code `OK | NOT OK`?

**Comment by sw-yx** [comment]
@frederikhors yes thats a good idea and would make the recipe more useful!!

**Comment by frederikhors** [comment]
@sw-yx I don't know where to start: hence my question for more clarity.

**Comment by sw-yx** [comment]
haha i dont know either. will leave this open in case anyone wants to tackle it - i honestlydont really know bc i dont use immutable enough
